File: packages/llm-obs-python-packages/python-shared/src/python_shared/http/resilience/tenant_partitioned_cache_store.py
import time
import threading
from typing import Dict, Any, Optional

class CacheEntry:
    def __init__(self, value: Any, expires_at: float):
        self.value = value
        self.expires_at = expires_at

class TenantPartitionedCacheStore:
    def __init__(self, max_partition_size: int = 100):
        self.max_partition_size = max_partition_size
        self._partitions: Dict[str, Dict[str, CacheEntry]] = {}
        self._lock = threading.RLock()

    def get(self, tenant_id: str, key: str) -> Optional[Any]:
        with self._lock:
            partition = self._partitions.get(tenant_id)
            if not partition:
                return None
            entry = partition.get(key)
            if not entry:
                return None
            if time.time() >= entry.expires_at:
                del partition[key]
                return None
            return entry.value

    def set(self, tenant_id: str, key: str, value: Any, ttl_ms: float = 5000.0) -> None:
        with self._lock:
            partition = self._partitions.setdefault(tenant_id, {})
            if len(partition) >= self.max_partition_size:
                oldest_key = next(iter(partition.keys()), None)
                if oldest_key:
                    del partition[oldest_key]
            expires_at = time.time() + (ttl_ms / 1000.0)
            partition[key] = CacheEntry(value, expires_at)

    def clear(self, tenant_id: str) -> None:
        with self._lock:
            self._partitions.pop(tenant_id, None)

    def clear_all(self) -> None:
        with self._lock:
            self._partitions.clear()

File: packages/llm-obs-node-packages/shared-infra/src/http/resilience/tenant-partitioned-cache-store.ts
/**
 * @file tenant-partitioned-cache-store.ts
 * @description Multi-Tenant Isolated Partitioned LRU Cache Store.
 * 
 * ALGORITHM & SPECIFICATION:
 * 1. Two-Level Partitioned Memory Map:
 *    - Structure: `Map<tenantId, Map<requestKey, CacheEntry>>`.
 *    - Guarantees zero cross-tenant data leaks.
 * 2. Invalidation Mechanics:
 *    - `clear(tenantId)`: Instantly invalidates all entries for the target tenant partition
 *      when mutating write RPCs (POST, PUT, PATCH, DELETE) occur.
 * 3. Bounded Partition LRU Capacity & TTL Eviction:
 *    - Each partition enforces `maxPartitionSize` (default: 100 entries) with Map LRU eviction.
 *    - Evicts items when `Date.now() >= expiresAt`.
 */

interface CacheEntry<T = unknown> {
  value: T;
  expiresAt: number;
}

export class TenantPartitionedCacheStore {
  private readonly partitions = new Map<string, Map<string, CacheEntry>>();

  constructor(private readonly maxPartitionSize = 100) {}

  public get<T>(tenantId: string, key: string): T | undefined {
    const partition = this.partitions.get(tenantId);
    if (!partition) return undefined;

    const entry = partition.get(key);
    if (!entry) return undefined;

    if (Date.now() >= entry.expiresAt) {
      partition.delete(key);
      return undefined;
    }

    return entry.value as T;
  }

  public set<T>(tenantId: string, key: string, value: T, ttlMs = 5000): void {
    let partition = this.partitions.get(tenantId);
    if (!partition) {
      partition = new Map<string, CacheEntry>();
      this.partitions.set(tenantId, partition);
    }

    if (partition.size >= this.maxPartitionSize) {
      const oldestKey = partition.keys().next().value;
      if (oldestKey) {
        partition.delete(oldestKey);
      }
    }

    partition.set(key, {
      value,
      expiresAt: Date.now() + ttlMs,
    });
  }

  public clear(tenantId: string): void {
    this.partitions.delete(tenantId);
  }

  public clearAll(): void {
    this.partitions.clear();
  }
}
